src/pip/_internal/network/auth.py

```diff
@@ -266,27 +265,57 @@ def _get_new_credentials(
                 logger.debug("Found credentials in index url for %s", netloc)
                 return index_url_user_password
 
-        # Get creds from netrc if we still don't have them
-        if allow_netrc:
-            netrc_auth = get_netrc_auth(original_url)
-            if netrc_auth:
-                logger.debug("Found credentials in netrc for %s", netloc)
-                return netrc_auth
-
-        # If we don't have a password and keyring is available, use it.
-        if allow_keyring:
-            # The index url is more specific than the netloc, so try it first
-            # fmt: off
-            kr_auth = (
-                get_keyring_auth(index_url, username) or
-                get_keyring_auth(netloc, username)
-            )
-            # fmt: on
+        netrc_auth = get_netrc_auth(original_url)
+        if netrc_auth is not None:
+            logger.debug("Found credentials in netrc for %s", netloc)
+            return netrc_auth
+
+        kr_auth = self._find_key_ring_credentials(
+            index_url, index_url_user_password, netloc, username
+        )
+        if kr_auth is not None:
+            return kr_auth
+
+        return username, password
+
+    def _find_key_ring_credentials(
```

Comment:
Order of getting keyring user:
1. Username from artifact url
2. Username from index url
3. Username from pip.conf (or command line argument)